improve-digital-htb.js

```diff
@@ -144,19 +149,37 @@ function ImproveDigitalHtb(configs) {
         var queryObj = {};
         var callbackId = System.generateUniqueId();
 
-        /* Change this to your bidder endpoint.*/
-        var baseUrl = Browser.getProtocol() + '//someAdapterEndpoint.com/bid';
-
-        /* ---------------- Craft bid request using the above returnParcels --------- */
-
-
-        /* -------------------------------------------------------------------------- */
-
-        return {
-            url: baseUrl,
-            data: queryObj,
-            callbackId: callbackId
-        };
+        var requestObject = {};
+
+        if(returnParcels && returnParcels.length) {
+            var requestParameters = {
+                returnObjType: __adServerClient.CONSTANTS.RETURN_OBJ_TYPE.DEFAULT,
+                libVersion: 'IX-' + __profile.version,
+                requestId: callbackId,
+                callback: 'window.' + SpaceCamp.NAMESPACE + '.' + __profile.namespace + '.adResponseCallback'
+            };
+            requestParameters.singleRequestMode = false;
+
+            var protocol = Browser.getProtocol();
+            requestParameters.secure = protocol === "https" ? __adServerClient.CONSTANTS.SECURE : __adServerClient.CONSTANTS.STANDARD;
+            var requestObject = [];
+            for (var parcelCounter = 0; parcelCounter < returnParcels.length; parcelCounter++) {
+                var parcel = returnParcels[parcelCounter];
+                if (parcel) {
+                    // The client needs an adUnitId.  The htSlot Id is guaranteed to be unique for the page
+                    var xSlotCopy = JSON.parse(JSON.stringify(parcel.xSlotRef));
+                    xSlotCopy.adUnitId = parcel.htSlot.getId();
+                    xSlotCopy.id = parcel.requestId;
+                    requestObject.push(xSlotCopy);
+                }
+            }
+            var request = __adServerClient.createRequest(requestObject, requestParameters);
+            if (request && request.requests && request.requests[0] && request.requests[0].url) {
+                queryObj.url = request.requests[0].url;
+                queryObj.callbackId = callbackId;
+            }
+        }
+        return queryObj;
```

Comment:
Please ensure the size is passed in the request and correct size bid response is returned from the ad server.

Reply:
As part of placement setup in our system, you define a list of sizes for that placement.  Where there is one size, the placement will only ever return that size.  Where there are multiple sizes, the placement will return any one of the sizes, depending on the dynamics of the auction.  Therefore, it is not necessary to include the size in the request to our ad servers.  The only time you should include a size in the request is if you have a placement with multiple sizes and you need the ad returned to be of one of those specific sizes.  I'll give examples using our 2 test placements.

Example 1 - Placement ID 1159201
This placement has been set up in our system with size 300x250.  When you call this placement, it always returns a 300x250 placement

Example 1 - Placement ID 1159200
This placement has been set up in our system woth sizes 728x90 and 970x250.  When it is called, it sometimes returns a 970x250 creative and sometimes returns a 728x90 creative.  We could, for example, specify w:728 and h:90 in the request to force it to always return 728x90 but in our tests we dont.

The two test placements do not currently have specific sizes in the requests because we don't want to do that for testing.

As for the responses, looking at the raw response, we always include the correct sizes:

Example 1 - Placement ID 1159201
Request: http://ad.360yield.com/hb?jsonp=%7B%22bid_request%22%3A%7B%22id%22%3A%22_1111111111%22%2C%22callback%22%3A%22window.headertag.ImproveDigitalHtb.adResponseCallback%22%2C%22version%22%3A%22IX-2.1.1-JS-5.1.1%22%2C%22imp%22%3A%5B%7B%22id%22%3A%22_2222222222%22%2C%22pid%22%3A1159200%2C%22banner%22%3A%7B%7D%7D%5D%7D%7D
Response: window.headertag.ImproveDigitalHtb.adResponseCallback({"id":"_1111111111","site_id":214190,"bid":[{"isNet":false,"id":"_2222222222","lid":292009,"advid":"5279","price":1.21481837863844,"nurl":"http://ad.360yield.com/imp_pixel?ic=eW8snwRv-qY.wESEg4YKnTlCiHgsVEvzd.c8z7VKcm2NXCc-ER66c7gB2xa98kas8GqYgCNcJFLGqi.RixRHiUCXupjFePhvOyZ1m5OM9SXriTZJXMIhx4yxlBm9Y9c9ZHQ1ImqArdzV1noBYNPEGWjNLKh9Mg2bq3HDfyZbLVjWk7aahNhbk-Qxb0y1ItJlIsIXBpI47yVE2NfQoWLuhy3Ez.zQptUjIOxuksnSWLE8p67yOQTmxqHdfOlyzJcm9-rLb-zkyVPS6ESA1eiiU6trMVyTPydFt8LkrWhUv9vSCQK0Du9lMLIiS-ofQCEpWmaDcp4MbWb4im-ii7up-Gnd9z6dJHyOQvZQIshJ-zTQXTekF3RIGcWMBe.pAEqXFTUwNLgWf8Zhc72h7w95k0ECzRM8IEaBxUYk4D483cBc3npn9mPYw2-sUB7Bzm47ZegvCJQlBcSWxlvXFuUxn374Iz49GXhFjoz9BS-q9.4kj0f7nnJDM9Vf35s0AnN1nOWaNisrAZQW5vN-JRNCQzidTn4qpEu-EseV4PxRiioy6VU=","h":250,"pid":1159200,"sync":["http://d5p.de17a.com/getuid/improve_digital?publisher_user_id=e39690a7-3174-49cb-9e65-3ae1cbe5473e&publisher_dsp_id=61&publisher_call_type=redirect&publisher_redirecturl=http://ad.360yield.com/match"],"crid":"445083","w":970,"cid":"108608","adm":"document.writeln(\"<a href=\\\"http:\\/\\/ad.360yield.com\\/click\\/eW8snzwRfloUcHx9MDduzLc2g9zqB6IxdM9H7zSdPWVrHqcwt0TPmu8QRiwGre3m7Ydl5bJcpuoj4livYRUJO52drMOPvVhlaIv3F20anGCi5fFBuTiv2R48Cxbu-Oxfi73FgVNl1E5GBsPbsGPrslb8rLC8CIQxRHGrrWSfXLyu0PATU.dnhiQ34mnLfXTTDkZZRZuaMQ0MRa3YmO0Qag0d3nwSLT4IXWttsTAE6uKPqouL7tiLQg-H3A7GcE9aV.QTGrCk4QHGwVcQkFEUxfm1wGePwhDLmvlcub0Rs6hPiR2Pk.Q.fVd4qqJ0tJoFWktAF1AZMRZkoQsu8BdlfRMbNZ35ZpaIPOU3aynJTIN-SMXPJ62zAlgPVXrPRMXHXsGAro5TxTkXQ7LTKtBQ5u8sss8XqE4sZGQzpmRLPSnMtz0wJqiqUU6HJZq7EHHCKHHCqa-9r5AgLAQx36t7PVxAU48vkdRbZ8tOs3Ai.sfNpmbj8Hqu0.P.o7ysdgHjC5R8.1JpqInafQFQ4PlsmNpzbkBDZQ6A8.Yms1WxOC0fwLU=\\/\\/http%3A%2F%2Fwww.improvedigital.com\\\" target=\\\"_blank\\\"><img style=\\\"border: 0;\\\" border=\\\"0\\\" width=\\\"970\\\" height=\\\"250\\\" src=\\\"http:\\/\\/creative.360yield.com\\/file\\/232040\\/ImpDig-970x250.jpg\\\" alt=\\\"\\\"\\/><\\/a>\");document.writeln(\"<improvedigital_ad_output_information tp_id=\\\"\\\" buyer_id=\\\"0\\\" rtb_advertiser=\\\"\\\" campaign_id=\\\"108608\\\" line_item_id=\\\"292009\\\" creative_id=\\\"445083\\\" crid=\\\"0\\\" placement_id=\\\"1159200\\\"><\\/improvedigital_ad_output_information>\");"}],"debug":""});

As you can see, we have "w" and "h" fields with the correct dimensions (970x250)

Refreshing this, we get the following Response:
window.headertag.ImproveDigitalHtb.adResponseCallback({"id":"_1111111111","site_id":214190,"bid":[{"isNet":false,"id":"_2222222222","lid":292009,"advid":"5279","price":1.21481837863844,"nurl":"http://ad.360yield.com/imp_pixel?ic=m2cFMQRtkuGiDjvHsZXR4ABIQXznlxYcUKA0vPQNCzDjRZOAazy-8cd64c8D0cC.RO.He69e8t6CdXODLwv4k1zL0WZ6-6u6VFcvedLwBEYG7RLEer-PQ9LWqLRGoTwonvfJvsMBnbJay8WADLPdQxfsotRP.DbKZJQXNttYq8WwgUuzHymXW7GSkc-JFBKB0I4tZ3lMfiHM5rQH3tyvxf2WegvzxWHpKxIyxZ9UGkGabBXJvIASh9KWazIw6b5k-o6sbaUqSTuEBrm4ybxCd3Bf5oSpHe10l9psulRw6.YO33RtNKZ0fiz6PD0kHZFDbMb7IBl7uyuoXB6xIuuUkvmvcA3xkY3PNfoXFgNNtJExfEJQNkPyCcz1hIxmRHLN4BmE3qCohT7O-AYNmadL.1IIX-YFeq2ikK-of9sj.vcdm4xtVu41pw.GOM7G-55gcw-F7.PkwEqNxfk9ckSYpyigq4Jkd0XxHOyZpN0Dkzx9.BvkHWqFSwJtmHyyygdBHdr8h1jTeRZBTG0QS3qC2htmYg==","h":90,"pid":1159200,"sync":["http://d5p.de17a.com/getuid/improve_digital?publisher_user_id=e39690a7-3174-49cb-9e65-3ae1cbe5473e&publisher_dsp_id=61&publisher_call_type=redirect&publisher_redirecturl=http://ad.360yield.com/match"],"crid":"445082","w":728,"cid":"108608","adm":"document.writeln(\"<a href=\\\"http:\\/\\/ad.360yield.com\\/click\\/m2cFMSnpsY0OdE.aclB1QkvDxBJ0kSMqzjf7HPd5HcSy.GygYQE.qNSUY2-4EHKfRrGTPzJ.MfljkbMau0gZPKp0ThPZsk1rdYJyjGM2p6XlqcrKEPPmXDAPODXce.v-Pxl-tDJjdC2deEfuzuGXnMNcfYFAtK7R.Y0m416NTK8y51hsl4tZUlqyHrBzDSxPH1pILZmb-rcimTdIcWQPnaiSxfkUHiP9h5zkaSsX4Yd38MiryH8jZUq.CO73HRp-nj3gRpMTW5eRfpDNyu7lKrY8TnRNjeryZaVJEX7TBygvA8071NPYlGxF8N4jyPw7sM3kVXg-cbpcTsMOncqHj6cF9Hn1wG6lGKzb7mPqpzb.RxWeE-gnhkuc6fLVeOYoeMHNwO35uTEy.iFtu-AOCw7OHe3eacusAxAXH0qv1m-myPij-sUmkJKeb-SNsP3EEaQkE3Vg-EKkZXe0J84kOxRxLxZdhB8q0.v7n2zaBn6YidS06sr3qalJTSDjor.uL9Ct3BBAPdGmPjwUav891vD0Vg==\\/\\/http%3A%2F%2Fwww.improvedigital.com\\\" target=\\\"_blank\\\"><img style=\\\"border: 0;\\\" border=\\\"0\\\" width=\\\"728\\\" height=\\\"90\\\" src=\\\"http:\\/\\/creative.360yield.com\\/file\\/232039\\/ImpDig-728x90.gif\\\" alt=\\\"\\\"\\/><\\/a>\");document.writeln(\"<improvedigital_ad_output_information tp_id=\\\"\\\" buyer_id=\\\"0\\\" rtb_advertiser=\\\"\\\" campaign_id=\\\"108608\\\" line_item_id=\\\"292009\\\" creative_id=\\\"445082\\\" crid=\\\"0\\\" placement_id=\\\"1159200\\\"><\\/improvedigital_ad_output_information>\");"}],"debug":""});

As you can see, we have "w" and "h" fields with the correct dimensions (728x90)

We correctly populate it with the line of code: var bidSize = [curBid.w, curBid.h];

Can you please explain what the size problem is?